front/lib/api/assistant/visualization.ts

```diff
@@ -76,14 +82,36 @@ export async function getVisualizationPrompt({
 
   let prompt = visualizationSystemPrompt.trim() + "\n\n";
 
-  if (contentFragmentMessages.length > 0) {
+  const fileAttachments: string[] = [];
+  for (const m of conversation.content.flat(1)) {
+    if (isContentFragmentType(m)) {
```

Comment:
nit: You could create a `isUploadContentFragmentType(m)` and rely on `UploadedContentFragment`

Reply:
Just re-using what's above, but what would I gain from this  ?

Reply:
You can get the type of `m` to be correct in the block within the `if` if you write:
```ts
export function isUploadContentFragmentType(
  arg: MessageType
): arg is ContentFragmentType & UploadedContentFragment {
  return arg.type === "content_fragment" && arg.fileId != null;
}
```

(it's nit but cool)

Reply:
I also get proper narrowing by checking that file_id isn't null + it's a lot simpler to understand when people read it. 
If we were doing this exact same things in a lot of places then i'd agree though.

I love type guards as much as the next guy, but this one seems unnecessary to me